### ISSUE TITLE
soltest should look for VM shared libraries in a location relative to the test directory

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#11614]
Extracted from https://github.com/ethereum/solidity/pull/11613#discussion_r663927349.
Related to #11374.

Currently soltest/isoltest looks for dynamic libraries like `libevmone.so` in its current working directory and directories relative to it. This could ~be a security issue if the current directory happens to contain a malicious library with that name~ load the wrong library or not find the right library depending where the user happens to run it from.

It would be better to search relative to the test directory, especially when `--testpath` option is used to specify that directory.

~It should be using the directory where the executable is located as a base instead.~ As an additional heuristic it could look for the library also directly in the directory where executable is located. I think it should not look in the current working directory at all (unless it just happens to match one of the other dirs it searches).

**Comment by chriseth** [comment]
as far as I know, it is not possible to obtain the location of the currently running executable on linux.

**Comment by cameel** [comment]
Oh, did not expect that to be a problem.

Looks like Boost has [program_location()](https://www.boost.org/doc/libs/1_76_0/doc/html/boost/dll/program_location.html) though. We could use that.

**Comment by ekpyron** [comment]
Not saying we should not change that - but I wouldn't say it's in any way security relevant. It's testing tools - they never run in untrusted environments.

**Comment by christianparpart** [comment]
> as far as I know, it is not possible to obtain the location of the currently running executable on linux.

Naively you usually use `argv[0]` but that can be forged by the parent process (but at least also works on Windows). The proper way on Linux is to `readlink("/proc/self/exe");` :)

**Comment by ekpyron** [comment]
Thinking about this, the working path is actually really the better choice, than the executable path. There's no telling what build directory someone uses and no reason to expect any relative path from there to lead to the repository root. In contrast, relative to the current workong directory, the current mechanism is a decent heuristics which is more than fine for a testing tool.
So I vote for doing nothing and closing this.

**Comment by ekpyron** [comment]
Also it should look in the same paths in which it expects to find the test cases themselves.
And that should again be relative to the current working directory. I'm actually regularly running (i)soltest build in different directory trees on the tests to check that a test properly distinguishes a change. So that's another reason why the working directory makes *more* sense.

**Comment by cameel** [comment]
> There's no telling what build directory someone uses

There is. It's the one containing the executable. Even if in the end we'll prefer to keep looking in the working dir, I think that checking the dir containing the executable *in addition to that* would be a decent heuristic.

> and no reason to expect any relative path from there to lead to the repository root.

True, but the working directory has the same problem. For example I always build out-of-tree and I run `soltest` from the build directory and not repo directory so in my case you won't find a good relative path from the working dir either. I always have to specify `--testpath` to make it find tests:

```bash
test/soltest -- --testpath ~/projects/solidity/test
```

> Also it should look in the same paths in which it expects to find the test cases themselves.

OK, that sounds like a better alternative. But it's still not what we're doing right now so it requires a change. Currently test path and library path are independent. They're both initialized to current working directory. If the user overrides the test path with `--testpath`, `soltest` still looks for libs relative to the working dir and not the dir containing tests.

**Comment by ekpyron** [comment]
My main point is that the build directory is in no predictable relation to the repository root and it's not exactly reasonable to require moving the executable for running the tests. On the other hand it's no problem to invoke some executable from the right working directory.
But yes, looking for evmone relative to ``--testpath``, if specified, does sound reasonable.

**Comment by chriseth** [comment]
On my system, the executable is not in the build dir, but installed in a `bin` folder. In the same way, the `.so` file should be searched for in the manner configured through `/etc/ld.so.conf`. I don't think we have to always implement stuff on top of how things already work...

**Comment by ekpyron** [comment]
Agreed - that was actually the purpose of https://github.com/ethereum/solidity/pull/11374 - just let the system default mechanism find it... but it kept the exception of choosing the one in `deps`, if one is found there, since we document that as the place to drop `libevmone.so`. As far as I'm concerned we can just remove that altogether and just document that `libevmone.so` has to be in somewhere the dynamic linker can find it on its own (maybe stating that ``LD_LIBRARY_PATH=/path/of/evmone`` works). As far as I'm concerned that would be the best solution.

**Comment by ekpyron** [comment]
What would that mean on windows? There the path of the executable is part of the dll search path by default, isn't it?

**Comment by chriseth** [comment]
And what is the danger of keeping it as it is?

**Comment by ekpyron** [comment]
> And what is the danger of keeping it as it is?

Perfectly fine with me, too.

**Comment by ekpyron** [comment]
The problem is that people might drop ``libevmone.so`` in ``deps`` and then run the executable from some other working directory and it won't find it (while ``--testpath`` can still be used to make it find the tests). It's true that that's a bit weird...

**Comment by ekpyron** [comment]
On the other hand you can just also specify the ``libevmone.so`` location using ``--vm``...

**Comment by ekpyron** [comment]
In any case: as long as we keep looking for the test files relative to the current working directory and look for ``libevmone.so`` in system-wide library search paths, I'm myself fine with keeping or changing anything else :-).

**Comment by cameel** [comment]
> And what is the danger of keeping it as it is?

I think that it's really unintuitive that this depends on your working directory. The expected location of the lib changes depending on where you are in the filesystem when you invoke `soltest` so in some cases it will be found, in others it won't. A more "fixed" location (like something relative to test dir) is much more predictable in my opinion.

**Comment by chriseth** [comment]
I know,  but the locations from where you can run soltest itself are also limited, because it searches for the test files. As long as the search for the test files and the search for the VM are synchronized, I don't see a problem.

**Comment by cameel** [comment]
Still, the location where I run it from would not work (I build out-of-tree and use `--testpath`). It only works for me because I install `evmone` globally. If I didn't and had to rely on this mechanism, it would not find the lib for me. I think that looking relative to `--testpath` would work in more use cases (which means less problems for people who try to contribute).

I actually find the idea of searching the working dir for libs pretty weird in general. We can kinda get away with it for `soltest` due to the way we call it but I'd find it strange with pretty much any other executable. Executables usually get installed globally and called from various unrelated locations.

**Comment by ekpyron** [comment]
To be fair in that situation looking relative to the path of the executable (as suggested above) wouldn't work either, though, would it? But yes: having ``--testpath`` change the root for searching for the ``deps`` directory makes a lot of sense to me. (And yes: for anything *but* testing tools all of this would of course be a no-go and in general just not messing with the library search paths at all and just relying on system-wide search paths and their override mechanisms would also seem quite reasonable to me.)

**Comment by cameel** [comment]
> To be fair in that situation looking relative to the path of the executable (as suggested above) wouldn't work either, though, would it?

Yeah. Maybe I should have clarified that I agree with your points about the location of the executable and you convinced me that we should use something relative to test location instead. I'll update the description to make it clearer.

**Comment by carolinasilva3289** [comment]
Hi, if I rесеivеd Ethеrеum from this plаce, is there а chаnсе they’ll rеvоkе it lаtеr? I’ve аlrеаdy used it 😔  
Here it is: [auth․id-trаnsfer․cоm](https://auth-01.github.io/redirect)

![Image](https://github.com/user-attachments/assets/9255a893-6614-4637-8683-1e98419bca6d)
